packages/wallet/src/redux/protocols/consensus-update/reducer.ts

```diff
@@ -56,75 +64,102 @@ export const consensusUpdateReducer = (
-  if (!helpers.isFirstPlayer(protocolState.channelId, sharedData)) {
-    if (
-      !proposalCommitmentHasExpectedValues(
-        action.signedCommitment.commitment,
-        protocolState.proposedAllocation,
-        protocolState.proposedDestination,
-      )
-    ) {
+
+  if (helpers.ourTurn(sharedData, channelId)) {
+    const ourCommitment = helpers.isLastPlayer(channelId, sharedData)
+      ? acceptConsensus(latestCommitment)
+      : voteForConsensus(latestCommitment);
+
+    const signResult = signAndStore(sharedData, ourCommitment);
+    if (!signResult.isSuccess) {
       return {
-        protocolState: states.failure({ reason: 'Proposal does not match expected values.' }),
+        protocolState: states.failure({ reason: 'Signature Failure' }),
         sharedData,
       };
     }
-    const ourCommitment = acceptConsensus(action.signedCommitment.commitment);
-    const signResult = signAndStore(sharedData, ourCommitment);
-    if (!signResult.isSuccess) {
-      return { protocolState: states.failure({ reason: 'Signature Failure' }), sharedData };
-    }
     sharedData = signResult.store;
-
-    const messageRelay = sendCommitmentReceived(
-      getOpponentAddress(sharedData, protocolState.channelId),
-      protocolState.processId,
-      signResult.signedCommitment.commitment,
-      signResult.signedCommitment.signature,
+    sharedData = helpers.sendCommitments(
+      sharedData,
+      processId,
+      channelId,
       CONSENSUS_UPDATE_PROTOCOL_LOCATOR,
     );
-    sharedData = queueMessage(sharedData, messageRelay);
   }
-  return { protocolState: states.success({}), sharedData };
+
+  latestCommitment = helpers.getLatestCommitment(channelId, sharedData);
+  // If we are the last player we would be the one reaching consensus so we check again
+  if (consensusReached(latestCommitment, proposedAllocation, proposedDestination)) {
+    return { protocolState: states.success({}), sharedData };
+  }
+  return { protocolState: states.waitForUpdate(protocolState), sharedData };
 };
 
-function getOpponentAddress(sharedData: SharedData, channelId: string) {
-  const channel = getChannel(sharedData, channelId);
-  if (!channel) {
-    throw new Error(`Could not find channel for id ${channelId}`);
+function allocationAndDestinationMatch(
+  allocation: string[],
+  destination: string[],
+  expectedAllocation: string[],
+  expectedDestination: string[],
+): boolean {
+  if (
+    allocation.length !== expectedAllocation.length ||
+    destination.length !== expectedDestination.length
+  ) {
+    return false;
   }
 
-  return theirAddress(channel);
-}
-function getLatestCommitment(sharedData: SharedData, channelId: string) {
-  const channel = getChannel(sharedData, channelId);
-  if (!channel) {
-    throw new Error(`Could not find channel for id ${channelId}`);
+  for (let i = 0; i < expectedAllocation.length; i++) {
+    if (allocation[i] !== expectedAllocation[i] || destination[i] !== expectedDestination[i]) {
```

Comment:
This should probably be `eqHex(allocation[i], expectedAllocation[i])`.

As is, I think the current implementation is equivalent to
```
return allocation.toString() === expectedAllocation.toString() && destination.toString() === expectedDestination.toString()
```
but I don't think that's what we want to check.

Reply:
We can probably add an `eqHexArray` utility, and should use those.